From: Stefan Haberland <sth@linux.vnet.ibm.com>
Subject: s390/dasd: fix failing CUIR assignment under LPAR
Patch-mainline: v4.8-rc3
Git-commit: af7752106e4f12b4ee47b4eca3e7ba4bcec6e7e5
References: bnc#994430, LTC#144674

Description:  dasd: fix failing CUIR assignment under LPAR
Symptom:      After a control unit initiated reconfiguration CUIR the
              wrong channel path might be set offline/online.
Problem:      On LPAR the read message buffer command should be
              executed on the path it was received on otherwise there
              is a chance that the CUIR assignment might be faulty and
              the wrong channel path is set online/offline.
Solution:     Fix by setting the path mask accordingly.
              On z/VM we might not be able to do I/O on this path but
              there it does not matter on which path the read message
              buffer command is executed. Therefor implement a retry
              with an open path mask.
Reproduction: Start CUIR action on storage server with Linux on LPAR
              running. After some iterations the wrong path might be
              handled.

Upstream-Description:

              s390/dasd: fix failing CUIR assignment under LPAR

              On LPAR the read message buffer command should be executed on the path
              it was received on otherwise there is a chance that the CUIR assignment
              might be faulty and the wrong channel path is set online/offline.

              Fix by setting the path mask accordingly.
              On z/VM we might not be able to do I/O on this path but there it does
              not matter on which path the read message buffer command is executed.
              Therefor implement a retry with an open path mask.

              Signed-off-by: Stefan Haberland <sth@linux.vnet.ibm.com>


Signed-off-by: Stefan Haberland <sth@linux.vnet.ibm.com>
Acked-by: John Jolly <jjolly@suse.de>
---
 drivers/s390/block/dasd_eckd.c |   10 ++++++++++
 1 file changed, 10 insertions(+)

--- a/drivers/s390/block/dasd_eckd.c
+++ b/drivers/s390/block/dasd_eckd.c
@@ -4573,6 +4573,8 @@ static int dasd_eckd_read_message_buffer
 		return PTR_ERR(cqr);
 	}
 
+	cqr->lpm = lpum;
+retry:
 	cqr->startdev = device;
 	cqr->memdev = device;
 	cqr->block = NULL;
@@ -4617,6 +4619,14 @@ static int dasd_eckd_read_message_buffer
 			(prssdp + 1);
 		memcpy(messages, message_buf,
 		       sizeof(struct dasd_rssd_messages));
+	} else if (cqr->lpm) {
+		/*
+		 * on z/VM we might not be able to do I/O on the requested path
+		 * but instead we get the required information on any path
+		 * so retry with open path mask
+		 */
+		cqr->lpm = 0;
+		goto retry;
 	} else
 		DBF_EVENT_DEVID(DBF_WARNING, device->cdev,
 				"Reading messages failed with rc=%d\n"
